core/drlibc/drlibc.c

```diff
@@ -164,6 +155,32 @@ clear_icache(void *beg, void *end)
     __asm__ __volatile__("isb" : : : "memory");
 #    endif
 }
+
+void
+set_cache_line_size_using_ctr_el0(OUT size_t *dcache_line_size,
+                                  OUT size_t *icache_line_size)
+{
+#    ifndef DR_HOST_NOT_TARGET
+    static size_t cache_info = 0;
+
+    /* "Cache Type Register" contains:
+     * CTR_EL0 [31]    : 1
+     * CTR_EL0 [19:16] : Log2 of number of 4-byte words in smallest dcache line
+     * CTR_EL0 [3:0]   : Log2 of number of 4-byte words in smallest icache line
+     * https://developer.arm.com/docs/ddi0595/h/aarch64-system-registers/ctr_el0
+     *
+     * Also, the whitepaper below documents AArch64 words being 32 bits wide.
+     * https://developer.arm.com/-/media/Files/pdf/
+     * graphics-and-multimedia/Porting%20to%20ARM%2064-bit.pdf
+     */
+    if (cache_info == 0)
+        __asm__ __volatile__("mrs %0, ctr_el0" : "=r"(cache_info));
```

Comment:
Hmm, two issues with the static var if it's not called at init time: racy; will crash if .data is read-only.  Best to have a comment stating that this is required to be called at init time when linked into DR.

Reply:
Added the comment.

Reply:
As the variable was marked static even before (when it was part of clear_icache), I assume this condition is already satisfied.